runtime/compiler/optimizer/EscapeAnalysis.cpp

```diff
@@ -1344,6 +1346,46 @@ int32_t TR_EscapeAnalysis::performAnalysisOnce()
 
             if (candidate->escapesInColdBlocks())
                {
+               if (_initializedHeapifiedTemps == NULL)
```

Comment:
Yes, this was intentional.  The idea was that once we find we have a candidate with cold block escapes, we check what is already initialized in the entry block - autos that could potentially hold a reference to a candidate that is being heapified might have been initialized in the entry block by an earlier pass of Escape Analysis or for some other reason.

Then, as we generate heapification code for any candidate (not just the first), we update `_initializedHeapifiedTemps` to include any additional temps that could hold references to heapified objects, and generate initialization code in the method entry block if it's the first time that temporary has been encountered.

Unless I'm mistaken, the same temporaries could contain references to more than one heapified candidate, so I wanted to avoid redundantly initializing the same temporaries in such a case, in addition to avoiding redundantly initializing a temporary associated with a candidate that is heapified in more than one place.

I could add some further comments describing this if you think it would forestall questions for future readers of this code.

I'll also mention that this initial pass through the method entry block will set entries in `_initializedHeapifiedTemps` that are not relevant for the purposes of heapification - any auto that's initialized in the entry block will be included.  I could refine this by taking the union of the symrefs that are associated with the candidates that escape in cold blocks, and intersect that with the symrefs that are initialized in the entry block to find only the set of symrefs that are important for heapification that are initialized in the method entry block.

Reply:
Thanks Henry, yes some comments explaining the reasoning would help.

Reply:
By "initializing" in the above explanation, you mean "null initializing" probably at least for the autos that are associated with candidates that escape in cold blocks, is that correct ?

Reply:
> By "initializing" in the above explanation, you mean "null initializing" probably at least for the autos that are associated with candidates that escape in cold blocks

I meant any store to any auto - I was assuming that if some reference, either null or non-null was being assigned to an auto in the entry block, that would suffice for ensuring that the auto contained a value that would allow GC to function correctly, and that an explicit null initialization would be redundant.  If you feel it would be safest to only look for null initializations in the entry block, I will make that change.

Now that I've thought about this a little further, I realize that the code that I wrote takes into account that an exception might be thrown in the entry block, and stops checking for stores to autos at that point, but it doesn't take into account that there might be a GC point in the entry block.  A potential GC point should also stop the code from considering any further initialization.  I will correct that.

Reply:
Vijay @vijaysun-omr, based on our conversation in this review comment, I've added another commit 5eee94ccd6b9d557582be0421c10ceb59a2895f1.  This adds some additional comments describing how `_initializedHeapifiedTemps` is only set up when the first candidate with cold block escapes is encountered, checks for GC points in the entry block, and only considers stores to autos of `aconst(0)` as initializations in setting up `_initializedHeapifiedTemps`.

If you're OK with that change, I will squash down this commit and my earlier commit 353b298f097900d733cbd4f10a87f4ded688f3db.